blockchain/chain_sync/src/sync.rs

```diff
@@ -0,0 +1,493 @@
+                return_set.push(ts);
+                continue;
+            }
+
+            const REQUEST_WINDOW: u64 = 100;
+            let epoch_diff = u64::from(cur_ts.epoch() - to_epoch);
+            let _window = min(epoch_diff, REQUEST_WINDOW);
+
+            // // Load blocks from network using blocksync
+            // TODO add sending blocksync req back (requires some channel for data back)
+            // let tipsets: Vec<Tipset> = self
+            //     .network
+            //     .get_headers(ts.parents(), window)
+            //     .await
+            //     .map_err(|e| Error::Other(e))?;
+            let tipsets: Vec<Tipset> = vec![];
+
+            // Loop through each tipset received from network
+            for ts in tipsets {
+                if ts.epoch() < to_epoch {
+                    // Break out of sync loop if epoch lower than to tipset
+                    // This should not be hit if response from server is correct
+                    break 'sync;
+                }
+                // Check Cids of blocks against bad block cache
+                self.validate_tipset_against_cache(&ts.key(), &accepted_blocks)?;
+
+                accepted_blocks.extend_from_slice(ts.cids());
+                // Add tipset to vector of tipsets to return
+                return_set.push(ts);
+            }
+        }
+
+        let last_ts = return_set
+            .last()
+            .ok_or_else(|| Error::Other("Return set should contain a tipset".to_owned()))?;
+
+        // Check if local chain was fork
+        if last_ts.key() != to.key() {
+            if last_ts.parents() == to.parents() {
+                // block received part of same tipset as best block
+                // This removes need to sync fork
+                return Ok(return_set);
+            }
+            // TODO add fork to return set
+            let _fork = self.sync_fork(&last_ts, &to).await?;
+        }
+
+        Ok(return_set)
+    }
+
+    fn validate_tipset_against_cache(
+        &mut self,
+        ts: &TipSetKeys,
+        accepted_blocks: &[Cid],
+    ) -> Result<(), Error> {
+        for cid in ts.cids() {
+            if let Some(reason) = self.bad_blocks.get(cid).cloned() {
+                for bh in accepted_blocks {
+                    self.bad_blocks
+                        .put(bh.clone(), format!("chain contained {}", cid));
+                }
+
+                return Err(Error::Other(format!(
+                    "Chain contained block marked as bad: {}, {}",
+                    cid, reason
+                )));
+            }
+        }
+        Ok(())
+    }
+
+    async fn sync_fork(&mut self, _head: &Tipset, _to: &Tipset) -> Result<Vec<Arc<Tipset>>, Error> {
+        // TODO sync fork until tipsets are equal or reaches genesis
+        todo!()
+    }
+
+    // Persists headers from tipset slice to chain store
+    fn persist_headers(&self, tipsets: &[Tipset]) -> Result<(), DBError> {
```

Comment:
This seems like a duplicate from `persist_headers` that exists in the ChainStore?

Reply:
it's just mapping over a vector of tipsets and calling that fn